modules: memfault: Add automatic sending of coredumps over LTE

simensrostad opened the pull request:
Add automatic sending of coredumps over LTE:
 - Add new file memfault_lte_coredump.c that uses LTE cereg and PDN to determine if the device is connected to the network.

   If connected, the layer will trigger sending of a stored coredump. Library is implemented with retry logic and backoff.

 - Add overlay file to SLM that enables Memfault features
 - Update Memfault sample to use the new coredump feature.
 - Add missing features to Memfault sample
 - Enable assertions by default in SLM, there is no reason why it should be enabled as long as there is space.
 
 Implemented according to state diagram found here:
 https://miro.com/app/board/uXjVLrI-A64=/

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 1
### Inputs:
#### Sources:
sdk-nrf: PR head: [99efc153e674f6dd50cea695ba59bfdb7bac01a7](https://github.com/nrfconnect/sdk-nrf/pull/20040/commits/99efc153e674f6dd50cea695ba59bfdb7bac01a7)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [99efc153e674f6dd50cea695ba59bfdb7bac01a7](https://github.com/nrfconnect/sdk-nrf/pull/20040/commits/99efc153e674f6dd50cea695ba59bfdb7bac01a7)
> merge base: [1c36f9c6163f8b4f2b8d94693243b78da5465fdd](https://github.com/nrfconnect/sdk-nrf/commit/1c36f9c6163f8b4f2b8d94693243b78da5465fdd)
> target head (main): [b701a885a07e4e559d144eb80c467ccf4bc0a1e4](https://github.com/nrfconnect/sdk-nrf/commit/b701a885a07e4e559d144eb80c467ccf4bc0a1e4)
> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/1c36f9c6163f8b4f2b8d94693243b78da5465fdd..99efc153e674f6dd50cea695ba59bfdb7bac01a7)</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (14)</summary>

```
applications
│  ├── serial_lte_modem
│  │  ├── doc
│  │  │  │ slm_description.rst
│  │  ├── overlay-memfault.conf
│  │  ├── prj.conf
│  │  ├── src
│  │  │  │ slm_at_commands.c
modules
│  ├── memfault-firmware-sdk
│  │  ├── CMakeLists.txt
│  │  ├── Kconfig
│  │  │ memfault_lte_coredump.c
samples
│  ├── debug
│  │  ├── memfault
│  │  │  ├── boards
│  │  │  │  ├── nrf9151dk_nrf9151_ns.conf
│  │  │  │  ├── nrf9160dk_nrf9160_ns.conf
│  │  │  │  ├── nrf9161dk_nrf9161_ns.conf
│  │  │  │  ├── thingy91_nrf9160_ns.conf
│  │  │  │  │ thingy91x_nrf9151_ns.conf
│  │  │  ├── prj.conf
│  │  │  ├── src
│  │  │  │  │ main.c
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :orange_circle: Toolchain
- :orange_circle: Build twister
- :orange_circle: Integration tests
   - :orange_circle: test-fw-nrfconnect-nrf-iot_serial_lte_modem

<details><summary>Disabled integration tests</summary>

- 
   - desktop52_verification
   - doc-internal
   - test_ble_nrf_config
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-ble_samples
   - test-fw-nrfconnect-boot
   - test-fw-nrfconnect-chip
   - test-fw-nrfconnect-fem
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_mosh
   - test-fw-nrfconnect-nrf-iot_positioning
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_thingy91
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-nrf_crypto
   - test-fw-nrfconnect-ps
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-rs
   - test-fw-nrfconnect-tfm
   - test-fw-nrfconnect-thread
   - test-fw-nrfconnect-zigbee
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-find-my
   - test-sdk-mcuboot
   - test-sdk-pmic-samples
   - test-sdk-sidewalk
   - test-sdk-wifi
   - test-secdom-samples-public
</details>



*Note: This message is automatically posted and updated by the CI*

jtguggedal commented on modules/memfault-firmware-sdk/memfault_lte_coredump.c:
```diff
@@ -0,0 +1,443 @@
+
+	LOG_ERR("state_waiting_for_network_connection_run");
+
+	state_object->retry_count = 0;
```
Can be moved to `state_network_connected_entry()`

jtguggedal commented on modules/memfault-firmware-sdk/memfault_lte_coredump.c:
```diff
@@ -0,0 +1,443 @@
+	state_object->retry_count++;
+
+	if (state_object->retry_count >= CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRIES_MAX) {
+		event_send(EVENT_COREDUMP_SEND_RETRY_COUNT_EXCEEDED);
```
Could set state to finished directly here

peknis commented:
Worth a changelog entry?

MarkusLassila commented on applications/serial_lte_modem/doc/slm_description.rst:
```diff
@@ -336,6 +336,8 @@ The following configuration files are provided:
   It can be customized to fit your configuration (UART, baud rate, and so on).
   By default, it sets the baud rate of the PPP UART to 1 000 000.
 
+* :file:`overlay-memfault.conf` - Configuration file that enables `Memfault`_.
```
This needs a better description, on what kind of scenarios you would enable this overlay.

As far as I understand, this would work in a way that you encounter a crash, then you flash the SLM with memfault-overlay, which sends the stored crash. In any case this flow should be described.

MarkusLassila commented:
> > Has this changed? If yes, then the document should be updated. If not, then the use-cases for this would be quite limited.
> 
> I think it's still true that faults originating in secure domain will not be forwarded to Memfault. `CONFIG_TFM_ALLOW_NON_SECURE_FAULT_HANDLING` allows handling of non-secure faults only, as far as I understand, but you know a lot more about this topic than me, Markus 😬

Ah, you are correct. It mentions the secure regions. I was looking at this through lenses of having attempted to collect a crash dump with memfault with TF-M (for non-secure), in which I was ultimately not successful, but I take your word that it should work for NS.

peknis commented on applications/serial_lte_modem/doc/slm_description.rst:
```diff
@@ -336,6 +336,8 @@ The following configuration files are provided:
   It can be customized to fit your configuration (UART, baud rate, and so on).
   By default, it sets the baud rate of the PPP UART to 1 000 000.
 
+* :file:`overlay-memfault.conf` - Configuration file that enables `Memfault`_, for more information about the Memfault integration in NCS see :ref:`mod_memfault`.
```
```suggestion
* :file:`overlay-memfault.conf` - Configuration file that enables `Memfault`_.
  For more information about the Memfault integration in |NCS|, see :ref:`mod_memfault`.
```

simensrostad commented:
@MarkusLassila I've added more information about the new automatic coredump send feature. However, I have not included details on how Memfault stores and handles coredumps, as that falls outside the scope of this PR.

I've updated the changelog for SLM and the Memfault integration.
@peknis Could you please have a look?

simensrostad commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,17 @@ To have these configuration files in the include path, add the following to the
 
 .. memfault_config_files_end
 
+Automatic Sending of Coredumps to Memfault
```
I think its fine to leave it where it is.

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -227,6 +227,8 @@ nRF Machine Learning (Edge Impulse)
 Serial LTE modem
 ----------------
 
+* Added overlay :file:`overlay-memfault.conf` to enable Memfault.
```
```suggestion
* Added an overlay :file:`overlay-memfault.conf` file to enable Memfault.
```

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -615,7 +617,9 @@ Edge Impulse integration
 
-|no_changes_yet_note|
+* Added a new feature to automatically post coredumps to Memfault when network connectivity is available.
+  This feature can be enabled by setting the :kconfig:option:`MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` Kconfig option to ``y``.
```
```suggestion
  To enable this feature, set the :kconfig:option:`MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` Kconfig option to ``y``.
```

divipillai commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,19 @@ To have these configuration files in the include path, add the following to the
+
+This feature is useful when you want to post the coredump as soon as possible after a crash and it is not desirable to wait for the next periodic upload set by :kconfig:option:`CONFIG_MEMFAULT_HTTP_PERIODIC_UPLOAD_INTERVAL_SECS`.
+Alternatively, you can manually trigger the coredump upload by calling the function :c:func:`memfault_zephyr_port_post_data`.
+The :c:function:`memfault_coredump_has_valid_coredump` function can be used to check if a coredump is available.
```
```suggestion
The :c:func:`memfault_coredump_has_valid_coredump` function can be used to check if a coredump is available.
```

To fix the doc build error.

divipillai commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -615,7 +617,9 @@ Edge Impulse integration
 --------------------
 
-|no_changes_yet_note|
+* Added a new feature to automatically post coredumps to Memfault when network connectivity is available.
```
The entry could be moved to the library section - https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/releases_and_maturity/releases/release-notes-changelog.html#debug_libraries as the feature is mentioned there.

simensrostad replied:
Not sure, this kconfig is a part of NCSs integration with Memfault. It only adds a library to the existing suite of c files that we use to integrate Memfault.

MarkusLassila reviewed (approved):
Approving SLM parts

divipillai commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -615,7 +617,9 @@ Edge Impulse integration
 
-|no_changes_yet_note|
+* Added a new feature to automatically post coredumps to Memfault when network connectivity is available.
+  To enable this feature, set the :kconfig:option:`MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` Kconfig option to ``y``.
```
```suggestion
  To enable this feature, set the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` Kconfig option to ``y``.
```

nordicjm commented on modules/memfault-firmware-sdk/CMakeLists.txt:
```diff
@@ -31,6 +31,10 @@ zephyr_library_sources_ifdef(
         CONFIG_MEMFAULT_NCS_BT_METRICS
         memfault_bt_metrics.c)
 
+zephyr_library_sources_ifdef(
+        CONFIG_MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED
+        memfault_lte_coredump.c)
```
cmake indent is 2 spaces and the end bracket just randomly changes above, please open another pr fixing the style in this file

simensrostad replied:
Styling aligned with the rest of the file, agree it should be addressed in a separate PR

nordicjm commented on modules/memfault-firmware-sdk/Kconfig:
```diff
@@ -191,6 +191,33 @@ config MEMFAULT_NCS_IMPLEMENT_METRICS_COLLECTION
+	select SMF
+	select SMF_ANCESTOR_SUPPORT
+	select SMF_INITIAL_TRANSITION
+	help
```
add a depends on for the supported nrf91 soc Kconfigs

simensrostad replied:
Done

nordicjm reviewed (commented):
Commits need squashing

peknis commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,19 @@ To have these configuration files in the include path, add the following to the
+Automatic sending of coredumps to Memfault
+==========================================
+
+To post a stored coredump from a previous crash to Memfault upon network connection, set the option :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` to ``y``.
```
```suggestion
To post a stored coredump from a previous crash to Memfault upon network connection, set the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` Kconfig option to ``y``.
```

peknis commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,19 @@ To have these configuration files in the include path, add the following to the
+To post a stored coredump from a previous crash to Memfault upon network connection, set the option :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_ON_NETWORK_CONNECTED` to ``y``.
+The option is only supported for nRF91 Series devices.
+
+The library has built-in connection awareness and tries to post the coredump to a maximum of the number set in the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRIES_MAX` Kconfig option, in an interval of the time set in the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRY_INTERVAL_SECONDS` Kconfig option between each attempt.
```
```suggestion
The library has built-in connection awareness and tries to post the coredump to a maximum of the number set in the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRIES_MAX` Kconfig option, at an interval of the time set in the :kconfig:option:`CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRY_INTERVAL_SECONDS` Kconfig option between each attempt.
```

peknis commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,19 @@ To have these configuration files in the include path, add the following to the
+If at any point the network is lost during the retry process, the library waits for the device to reconnect before restarting the retry process.
+
+This feature is useful when you want to post the coredump as soon as possible after a crash and it is not desirable to wait for the next periodic upload set by :kconfig:option:`CONFIG_MEMFAULT_HTTP_PERIODIC_UPLOAD_INTERVAL_SECS`.
+Alternatively, you can manually trigger the coredump upload by calling the function :c:func:`memfault_zephyr_port_post_data`.
```
```suggestion
Alternatively, you can manually trigger the coredump upload by calling the :c:func:`memfault_zephyr_port_post_data` function.
```

peknis commented on doc/nrf/libraries/debug/memfault_ncs.rst:
```diff
@@ -58,6 +58,19 @@ To have these configuration files in the include path, add the following to the
+
+This feature is useful when you want to post the coredump as soon as possible after a crash and it is not desirable to wait for the next periodic upload set by :kconfig:option:`CONFIG_MEMFAULT_HTTP_PERIODIC_UPLOAD_INTERVAL_SECS`.
+Alternatively, you can manually trigger the coredump upload by calling the function :c:func:`memfault_zephyr_port_post_data`.
+The :c:func:`memfault_coredump_has_valid_coredump` function can be used to check if a coredump is available.
```
```suggestion
You can use the :c:func:`memfault_coredump_has_valid_coredump` function to check whether a coredump is available.
```

nordicjm commented on modules/memfault-firmware-sdk/Kconfig:
```diff
@@ -191,6 +191,43 @@ config MEMFAULT_NCS_IMPLEMENT_METRICS_COLLECTION
+	  Delay coredump posting to Memfault to avoid conflicts with the application's
+	  TLS connection setup that occurs typically LTE network is obtained.
+	  The delay duration is defined by CONFIG_MEMFAULT_NCS_POST_COREDUMP_RETRY_INTERVAL_SECONDS.
+#§
```
?

simensrostad replied:
[X]

rlubos commented:
@nordicjm Please revisit
